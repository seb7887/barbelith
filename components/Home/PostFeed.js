import React from 'react';
import Typography from '@material-ui/core/Typography';
import withStyles from '@material-ui/core/styles/withStyles';

import {
  getPostFeed,
  addPost,
  deletePost,
  likePost,
  unlikePost,
  addComment,
  deleteComment
} from '../../lib/api';

import NewPost from './NewPost';
import Post from './Post';

class PostFeed extends React.Component {
  state = {
    posts: [],
    text: '',
    image: '',
    isAddingPost: false,
    isDeletingPost: false
  };

  componentDidMount() {
    this.postData = new FormData();
    this.getPosts();
  }

  getPosts = () => {
    const { auth } = this.props;
    getPostFeed(auth.user._id)
      .then(posts => this.setState({ posts }));
  }

  handleChange = (e) => {
    e.preventDefault();
    let inputValue;

    // if target is an image
    if (e.target.name === 'image') {
      inputValue = e.target.files[0];
    } else {
      inputValue = e.target.value;
    }
    this.postData.set(e.target.name, inputValue);
    this.setState({ [e.target.name]: inputValue });
  }

  handleAddPost = () => {
    const { auth } = this.props;

    this.setState({ isAddingPost: true });
    addPost(auth.user._id, this.postData)
      .then(postData => {
        // add new post to the state's posts array
        const updatedPosts = [postData, ...this.state.posts];
        this.setState({
          posts: updatedPosts,
          isAddingPost: false,
          text: '',
          image: ''
        });
        this.postData.delete('image');
      })
      .catch(err => {
        this.setState({ isAddingPost: false });
      });
  }

  handleDeletePost = (deletedPost) => {
    this.setState({ isDeletingPost: true });
    deletePost(deletedPost._id)
      .then(postData => {
        const postIndex = this.state.posts.findIndex(
          post => post._id === postData._id
        );
        const updatedPosts = [
          ...this.state.posts.slice(0, postIndex),
          ...this.state.posts.slice(postIndex + 1)
        ];
        this.setState({
          posts: updatedPosts,
          isDeletingPost: false
        });
      })
      .catch(err => {
        this.setState({ isDeletingPost: false });
      });
  }

  handleToggleLike = (post) => {
    const { auth } = this.props;
    const isPostLiked = post.likes.includes(auth.user._id);
    const sendRequest = isPostLiked ? unlikePost : likePost;
    sendRequest(post._id)
      .then(postData => {
        const postIndex = this.state.posts.findIndex(
          post => post._id === postData._id
        );
        const updatedPosts = [
          ...this.state.posts.slice(0, postIndex),
          postData,
          ...this.state.posts.slice(postIndex + 1)
        ];
        this.setState({ posts: updatedPosts });
      })
      .catch(err => console.log(err));
  }

  handleAddComment = (postId, text) => {
    const comment = { text };
    addComment(postId, comment)
      .then(postData => {
        const postIndex = this.state.posts.findIndex(
          post => post._id === postData._id
        );
        const updatedPosts = [
          ...this.state.posts.slice(0, postIndex),
          postData,
          ...this.state.posts.slice(postIndex + 1)
        ];
        this.setState({ posts: updatedPosts });
      })
      .catch(err => console.log(err));
  }

  handleDeleteComment = (postId, comment) => {
    deleteComment(postId, comment)
      .then(postData => {
        const postIndex = this.state.posts.findIndex(
          post => post._id === postData._id
        );
        const updatedPosts = [
          ...this.state.posts.slice(0, postIndex),
          postData,
          ...this.state.posts.slice(postIndex + 1)
        ];
        this.setState({ posts: updatedPosts });
      })
      .catch(err => console.log(err));
  }

  render() {
    const { classes, auth } = this.props;
    const { posts, text, image, isAddingPost, isDeletingPost } = this.state;

    return (
      <div className={classes.root}>
        <Typography
          variant='h4'
          component='h1'
          align='center'
          color='secondary'
          className={classes.title}
        >
          Timeline
        </Typography>
        <NewPost
          auth={auth}
          text={text}
          image={image}
          isAddingPost={isAddingPost}
          handleChange={this.handleChange}
          handleAddPost={this.handleAddPost}
        />
        {posts.map(post => (
          <Post
            key={post._id}
            auth={auth}
            post={post}
            isDeletingPost={isDeletingPost}
            handleDeletePost={this.handleDeletePost}
            handleToggleLike={this.handleToggleLike}
            handleAddComment={this.handleAddComment}
            handleDeleteComment={this.handleDeleteComment}
          />
        ))}
      </div>
    );
  }
}

const styles = theme => ({
  root: {
    paddingBottom: theme.spacing.unit * 2
  },
  title: {
    padding: theme.spacing.unit * 2
  }
});

export default withStyles(styles)(PostFeed);
