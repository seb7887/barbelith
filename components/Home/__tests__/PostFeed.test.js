/* eslint-env jest */
import 'react-testing-library/cleanup-after-each';
import 'jest-dom/extend-expect';

import { render, fireEvent, waitForElement } from 'react-testing-library';
import PostFeed from '../PostFeed';

describe('<PostFeed/>', () => {
  it('', () => {
    expect(1).toBe(1);
  });
});
