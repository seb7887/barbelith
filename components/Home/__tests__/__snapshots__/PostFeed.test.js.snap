// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostFeed/> renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="PostFeed-root-1"
  >
    <h1
      class="MuiTypography-root-3 MuiTypography-h4-18 MuiTypography-colorSecondary-34 MuiTypography-alignCenter-26 PostFeed-title-2"
    >
      Timeline
    </h1>
    <div
      class="MuiPaper-root-44 MuiPaper-elevation1-47 MuiPaper-rounded-45 MuiCard-root-43 NewPost-card-39"
      data-testid="card"
    >
      <div
        class="MuiCardHeader-root-71"
      >
        <div
          class="MuiCardHeader-avatar-72"
        >
          <div
            class="MuiAvatar-root-77"
          >
            <img
              class="MuiAvatar-img-79"
              src="test.jpg"
            />
          </div>
        </div>
        <div
          class="MuiCardHeader-content-74"
        >
          <h2
            class="MuiTypography-root-3 MuiTypography-h6-20"
          >
            test
          </h2>
        </div>
      </div>
      <div
        class="MuiCardContent-root-80 NewPost-cardContent-40"
      >
        <div
          class="MuiFormControl-root-81 MuiFormControl-marginNormal-82 MuiFormControl-fullWidth-84"
          data-testid="textfield"
          row="2"
        >
          <label
            class="MuiFormLabel-root-96 MuiInputLabel-root-85 MuiInputLabel-formControl-90 MuiInputLabel-animated-93 MuiInputLabel-shrink-92 MuiInputLabel-outlined-95"
            data-shrink="true"
          >
            Add a status
          </label>
          <div
            class="MuiInputBase-root-116 MuiOutlinedInput-root-103 MuiInputBase-fullWidth-125 MuiInputBase-formControl-117 MuiInputBase-multiline-124 MuiOutlinedInput-multiline-109"
          >
            <fieldset
              aria-hidden="true"
              class="MuiPrivateNotchedOutline-root-133 MuiOutlinedInput-notchedOutline-110"
              style="padding-left: 8px;"
            >
              <legend
                class="MuiPrivateNotchedOutline-legend-134"
                style="width: 0px;"
              >
                <span>
                  â€‹
                </span>
              </legend>
            </fieldset>
            <div
              class="MuiPrivateTextarea-root-135"
            >
              <textarea
                aria-hidden="true"
                class="MuiPrivateTextarea-textarea-136 MuiPrivateTextarea-shadow-137"
                readonly=""
                rows="1"
                tabindex="-1"
              />
              <textarea
                aria-hidden="true"
                class="MuiPrivateTextarea-textarea-136 MuiPrivateTextarea-shadow-137"
                readonly=""
                rows="1"
                tabindex="-1"
              />
              <textarea
                aria-invalid="false"
                class="MuiPrivateTextarea-textarea-136 MuiInputBase-input-126 MuiOutlinedInput-input-111 MuiInputBase-inputMultiline-128 MuiOutlinedInput-inputMultiline-113"
                name="text"
                placeholder="What's on your mind, test?"
                rows="1"
                style="height: 19px;"
              />
            </div>
          </div>
        </div>
        <input
          accept="image/*"
          class="NewPost-input-41"
          data-testid="image"
          id="image"
          name="image"
          type="file"
        />
        <label
          for="image"
        >
          <span
            class="MuiButtonBase-root-144 MuiIconButton-root-138 MuiIconButton-colorSecondary-141"
            role="button"
            tabindex="0"
          >
            <span
              class="MuiIconButton-label-143"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-147"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <defs>
                  <path
                    d="M24 24H0V0h24v24z"
                    id="a"
                  />
                </defs>
                <path
                  d="M3 4V1h2v3h3v2H5v3H3V6H0V4h3zm3 6V7h3V4h7l1.83 2H21c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H5c-1.1 0-2-.9-2-2V10h3zm7 9c2.76 0 5-2.24 5-5s-2.24-5-5-5-5 2.24-5 5 2.24 5 5 5zm-3.2-5c0 1.77 1.43 3.2 3.2 3.2s3.2-1.43 3.2-3.2-1.43-3.2-3.2-3.2-3.2 1.43-3.2 3.2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root-185"
            />
          </span>
        </label>
        <span />
      </div>
      <div
        class="MuiCardActions-root-156 NewPost-cardActions-42"
      >
        <button
          class="MuiButtonBase-root-144 MuiButtonBase-disabled-145 MuiButton-root-159 MuiButton-contained-170 MuiButton-containedSecondary-172 MuiButton-raised-173 MuiButton-raisedSecondary-175 MuiButton-disabled-179 MuiCardActions-action-158"
          data-testid="submit"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label-160"
          >
            Post
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
